internal/agent/runtime/docker.go

```diff
@@ -0,0 +1,204 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+
+	retry "github.com/avast/retry-go"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/go-logr/logr"
+	"github.com/tinkerbell/tink/internal/agent"
+	"github.com/tinkerbell/tink/internal/agent/runtime/internal"
+	"github.com/tinkerbell/tink/internal/agent/workflow"
+	"github.com/tinkerbell/tink/internal/ptr"
+	"k8s.io/apimachinery/pkg/util/rand"
+)
+
+var _ agent.ContainerRuntime = &Docker{}
+
+// Docker is a docker runtime that satisfies agent.ContainerRuntime.
+type Docker struct {
+	log    logr.Logger
+	client *client.Client
+}
+
+// Run satisfies agent.ContainerRuntime.
+func (d *Docker) Run(ctx context.Context, a workflow.Action) error {
+	pullImage := func() error {
+		// We need the image to be available before we can create a container.
+		image, err := d.client.ImagePull(ctx, a.Image, types.ImagePullOptions{})
+		if err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+		defer image.Close()
+
+		// Docker requires everything to be read from the images ReadCloser for the image to actually
+		// be pulled. We may want to log image pulls in a circular buffer somewhere for debugability.
+		if _, err = io.Copy(io.Discard, image); err != nil {
+			return fmt.Errorf("docker: %w", err)
+		}
+
+		return nil
+	}
+
+	err := retry.Do(pullImage, retry.Attempts(5), retry.DelayType(retry.BackOffDelay))
+	if err != nil {
+		return err
+	}
+
+	// TODO: Support all the other things on the action such as volumes.
+	cfg := container.Config{
+		Image: a.Image,
+		Env:   toDockerEnv(a.Env),
```

Comment:
we might want to set the failure file locations to be env vars for all containers. that way actions dont need to "know" where this location is and we can change it if needed. action writers just look up the env var and get the files.